src/builtins.c

```diff
@@ -893,6 +893,11 @@ static inline size_t get_checked_fieldindex(const char *name, jl_datatype_t *st,
         if (idx >= jl_datatype_nfields(st))
             jl_bounds_error(v, arg);
     }
+    else if (jl_isa(arg, (jl_value_t*)jl_number_type)) {
+        // This always throws but makes the error way better
+        // Otherwise it would think you were trying to pass a symbol which isn't what you were trying to do.
+        JL_TYPECHKS(name, long, arg);
+    }
```

Comment:
What about:
```c
    if (jl_is_long(arg)) {
        idx = jl_unbox_long(arg) - 1;
        if (idx >= jl_datatype_nfields(st))
            jl_bounds_error(v, arg);
    }
    else if (jl_is_symbol(arg)) {
        idx = jl_field_index(st, (jl_sym_t*)arg, 1);
    }
    else {
        jl_value_t *ts[2] = {(jl_value_t*)jl_long_type, (jl_value_t*)jl_symbol_type};
        jl_value_t *t = jl_type_union(ts, 2);
        jl_type_error("getfield", t, arg);
    }
```

It seems a bit awkward to use `jl_isa` check against `jl_number_type` here (since the `Number` type is not the only possible wrong choice).

Reply:
```julia
julia> getfield(Ref("foo"), "s")
ERROR: TypeError: in getfield, expected Union{Int64, Symbol}, got a value of type String
Stacktrace:
 [1] top-level scope
   @ REPL[1]:1
```

Reply:
my thought with this implementation is that Numbers are the only things that might be Integers so if they pass a Number, they probably wanted an Int, and otherwise, they probably wanted a symbol.

Reply:
It's also possible to call `getfield(x, :true)` when meaning `getfield(x, Symbol("true"))`.